### PR TITLE
Split to sentences implementation

### PR DESCRIPTION
**Pull request by maxkoryukov**
Hello!

This PR contains the implementation of Sentence Buffer: Split

Usage:
```shell
./ccextractor -sbs ~/source.ts
```

Currently, it works only with `sub->type == CC_BITMAP`. Implementation details - in comments to the PR.

#### New output

```
1
00:00:00,001 --> 00:00:00,189
Oleon costs.

2
00:00:00,191 --> 00:00:00,783
buried in the annex, 95 Oleon costs.

3
00:00:00,785 --> 00:00:05,159
Didn't want to acknowledge the pressures on hospitals, schools and infrastructure.
```

#### Old output

```
1
00:00:00,001 --> 00:00:00,000
Oleon

2
00:00:00,001 --> 00:00:00,189
Oleon costs.

3
00:00:00,190 --> 00:00:00,889
buried in the annex, 95 Oleon costs.
Didn't

4
00:00:00,890 --> 00:00:01,129
buried in the annex, 95 Oleon costs.
Didn't want

5
00:00:01,130 --> 00:00:01,359
buried in the annex, 95 Oleon costs.
Didn't want to

6
00:00:01,360 --> 00:00:02,059
buried in the annex, 95 Oleon costs.
Didn't want to acknowledge

7
00:00:02,060 --> 00:00:02,299
buried in the annex, 95 Oleon costs.
Didn't want to acknowledge the

8
00:00:02,300 --> 00:00:03,419
Didn't want to acknowledge the
pressures

9
00:00:03,420 --> 00:00:03,609
Didn't want to acknowledge the
pressures on

10
00:00:03,610 --> 00:00:04,029
Didn't want to acknowledge the
pressures on hospitals,

11
00:00:04,030 --> 00:00:04,779
Didn't want to acknowledge the
pressures on hospitals, schools

12
00:00:04,780 --> 00:00:05,019
Didn't want to acknowledge the
pressures on hospitals, schools and

13
00:00:05,020 --> 00:00:05,159
pressures on hospitals, schools and
infrastructure.

```

### CODE REVIEW

**Comment by maxkoryukov** [comment]
fix maxkoryukov/ccextractor#1

File: src/lib_ccx/ccx_encoders_common.c [hunk 1]
```diff
@@ -1045,203 +1041,204 @@ int encode_sub(struct encoder_ctx *context, struct cc_subtitle *sub)
...
 		// in sentences
 		if (sub->type == CC_BITMAP)
-			wrote_something = write_cc_bitmap_to_sentence_buffer(sub, context);
+			sub = reformat_cc_bitmap_through_sentence_buffer(sub, context);
+
+		if (NULL==sub)
+			return wrote_something;
 	}
-	else
+	// Write subtitles as they come
```
Comment (maxkoryukov) [inline src/lib_ccx/ccx_encoders_common.c]: There is a lot of changes lines. BUT, the most changes - deleted <TAB> symbol. Just ignore whitespaces here:

https://github.com/CCExtractor/ccextractor/pull/491/files?w=1

File: src/lib_ccx/ccx_encoders_common.c [hunk 2]
```diff
@@ -1045,203 +1041,204 @@ int encode_sub(struct encoder_ctx *context, struct cc_subtitle *sub)
 		// Write to a buffer that is later s+plit to generate split
 		// in sentences
 		if (sub->type == CC_BITMAP)
-			wrote_something = write_cc_bitmap_to_sentence_buffer(sub, context);
+			sub = reformat_cc_bitmap_through_sentence_buffer(sub, context);
```
Comment (maxkoryukov) [inline src/lib_ccx/ccx_encoders_common.c]: This is the most important point in this PR.

SBS works as **transformation filter**. It takes incoming subs, and convert them to other subs with `sub->type = CC_TEXT`. This transformation occurred before all other encoders, so other encoders remain unchanged.

File: src/lib_ccx/debug_def.h [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#ifndef _DEBUG_DEF_H_
```
Comment (maxkoryukov) [inline src/lib_ccx/debug_def.h]: This is just a helper for debugging. You could remove this file, and all references to `LOG_DEBUG`... But they are useful for debugging (with existing tests)

File: tests/Makefile [hunk 4]
```diff
@@ -0,0 +1,59 @@
+SHELL = /bin/sh
```
Comment (maxkoryukov) [inline tests/Makefile]: The new `tests` folder contains unit-tests for SBS. They are written with [libcheck](https://github.com/libcheck/check)

There is `README.md`, with short how-to instructions.

**Comment by maxkoryukov** [comment]
@cfsmp3, @canihavesomecoffee , is there a chance to merge this to the upstream?

**Comment by cfsmp3** [comment]
Not for 0.83 (which will be released today). Could happen for 0.84. Note
though that most likely you PR won't merge now since there's been many
changes, so you may need to resubmit.

**Pull request closed without merging**